nixos/roles/router/kresd/default.nix

```diff
@@ -0,0 +1,167 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}:
+let
+  fclib = config.fclib;
+  role = config.flyingcircus.roles.router;
+
+  # make additional lua packages available at runtime
+  knotPackage =
+    let
+      wrapped = pkgs.knot-resolver.override { extraFeatures = true; };
+    in
+    wrapped.overrideAttrs (old: {
+      buildInputs = old.buildInputs ++ (with pkgs.luajitPackages; [ luaposix ]);
+    });
+in
+lib.mkIf role.enable {
+  # set by upstream kresd module
+  networking.resolvconf.useLocalResolver = fclib.mkPlatform false;
+
+  services.kresd = {
+    enable = true;
+    package = knotPackage;
+    listenPlain = [
+      "0.0.0.0:53"
+      "[::]:53"
+    ];
+    extraConfig = ''
+      -- basic config
+
+      modules = {
+        'serve_stale < cache',
+        'hints > iterate',
+        'prefill',
+        'view'
+      }
+
+      cache.size = 1024*MB
+
+    ''
+    + lib.optionalString (role.sourceAddressV4 != null) ''
+      net.outgoing_v4('${role.sourceAddressV4}')
+    ''
+    + lib.optionalString (role.sourceAddressV6 != null) ''
+      net.outgoing_v6('${role.sourceAddressV6}')
+    ''
+    + ''
+
+      -- resilience: prefill the root zone data
+      prefill.config({
+          ['.'] = {
+              url = 'https://www.internic.net/domain/root.zone',
+              interval = 86400,
+              ca_file = '/etc/pki/tls/certs/ca-bundle.crt',
+          }
+      })
+
+      -- resilience: prefer serving stale data over not serving at all.
+      -- rfc8767 recommends a stale serving timeout of 1800s with a client
+      -- query timeout of 2s. however, kresd has a minimum granularity of
+      -- 2s (KR_CONN_RTT_MAX), so we need to configure a higher client query
+      -- timeout than recommended in the rfc to ensure that clients wait for
+      -- long enough to trigger the stale serving behaviour. we match the
+      -- rfc for the stale serving timeout though so we get the benefits
+      -- if the granularity changes in the future.
+      serve_stale.timeout = 1800
+
+      -- if a backend is detected as down, let it stay down for a minute so we
+      -- can immediately respond to requests. however, this also means backends
+      -- may take a minute to recover
+      cache.ns_tout(60000)
+
+      -- access control: only answer queries from ourselves and own address
+      -- space
+      view:addr('127.0.0.0/8', policy.all(policy.PASS))
+      view:addr('::1/128', policy.all(policy.PASS))
+
+      ${lib.concatMapStringsSep "\n" (
```

Comment:
I was wondering whether this will cause unnecessary hard reloads. I was considering to put this into firewalling code instead if it does.

Reply:
Well, the short answer is yes, this will cause a restart (not a reload) if the list changes. It's possible that kresd might restart fast enough for this to not be a problem, though this is indeed something which has bitten us with bind in the past.

As for the firewalling... if we put the list of allowed IP ranges in the firewall, then changing the allowed IPs will cause a *firewall* reload, which on the routers usually causes the BGP sessions to go down and cause a failover due to the BFD getting interrupted. I presume by "put this into firewalling" you mean restrict access to DNS to only "downstream" interfaces, and e.g. dropping incoming DNS from the transfer/uplink interfaces?